File: src/engine/allergen-engine.ts
// Allergen detection engine — real logic, not mock data
// Uses keyword matching (Phase 1), upgradeable to ML model

import type { Allergen } from '@/types/ingredient';

const ALLERGEN_DATABASE: Omit<Allergen, 'detected' | 'matchedKeywords'>[] = [
  {
    name: 'Wheat/Gluten',
    keywords: ['wheat', 'gluten', 'flour', 'semolina', 'durum', 'spelt', 'kamut', 'farina', 'couscous', 'bulgur', 'seitan'],
    severity: 'high',
  },
  {
    name: 'Milk/Dairy',
    keywords: ['milk', 'cream', 'butter', 'cheese', 'whey', 'casein', 'lactose', 'yogurt', 'ghee', 'curd'],
    severity: 'high',
  },
  {
    name: 'Soy',
    keywords: ['soy', 'soya', 'soybean', 'edamame', 'tofu', 'tempeh', 'miso', 'soy lecithin', 'soy protein'],
    severity: 'high',
  },
  {
    name: 'Egg',
    keywords: ['egg', 'albumin', 'globulin', 'lysozyme', 'mayonnaise', 'meringue', 'ovalbumin', 'ovomucin'],
    severity: 'high',
  },
  {
    name: 'Tree Nuts',
    keywords: ['almond', 'cashew', 'walnut', 'pecan', 'pistachio', 'macadamia', 'hazelnut', 'brazil nut', 'chestnut', 'pine nut'],
    severity: 'high',
  },
  {
    name: 'Peanut',
    keywords: ['peanut', 'groundnut', 'arachis'],
    severity: 'high',
  },
  {
    name: 'Fish',
    keywords: ['fish', 'cod', 'salmon', 'tuna', 'anchovy', 'sardine', 'tilapia', 'bass', 'trout'],
    severity: 'medium',
  },
  {
    name: 'Shellfish',
    keywords: ['shrimp', 'crab', 'lobster', 'clam', 'mussel', 'oyster', 'scallop', 'crawfish', 'prawn'],
    severity: 'medium',
  },
  {
    name: 'Sesame',
    keywords: ['sesame', 'tahini', 'halvah'],
    severity: 'medium',
  },
  {
    name: 'Sulfites',
    keywords: ['sulfite', 'sulphite', 'sulfur dioxide', 'sodium bisulfite', 'sodium metabisulfite', 'potassium bisulfite'],
    severity: 'low',
  },
];

export function detectAllergens(ingredientText: string): Allergen[] {
  const normalizedText = ingredientText.toLowerCase();

  return ALLERGEN_DATABASE.map((allergen) => {
    const matchedKeywords = allergen.keywords.filter((keyword) =>
      normalizedText.includes(keyword.toLowerCase())
    );

    return {
      ...allergen,
      detected: matchedKeywords.length > 0,
      matchedKeywords,
    };
  });
}

File: backend/app/services/allergen_engine.py
"""
Allergen Detection Engine
Detects 10 major allergen groups in ingredient text
"""

ALLERGEN_DATABASE = [
    {
        "name": "Wheat & Gluten",
        "keywords": ["wheat", "flour", "gluten", "semolina", "durum", "spelt", "kamut", "farro", "bulgur", "couscous", "seitan"],
        "severity": "high"
    },
    {
        "name": "Dairy & Milk",
        "keywords": ["milk", "dairy", "cheese", "butter", "cream", "yogurt", "whey", "casein", "lactose", "ghee", "paneer"],
        "severity": "high"
    },
    {
        "name": "Eggs",
        "keywords": ["egg", "albumin", "mayonnaise", "meringue", "lysozyme"],
        "severity": "high"
    },
    {
        "name": "Soy",
        "keywords": ["soy", "soya", "tofu", "tempeh", "edamame", "miso", "lecithin"],
        "severity": "medium"
    },
    {
        "name": "Tree Nuts",
        "keywords": ["almond", "cashew", "walnut", "pecan", "pistachio", "hazelnut", "macadamia", "brazil nut", "pine nut"],
        "severity": "high"
    },
    {
        "name": "Peanuts",
        "keywords": ["peanut", "groundnut", "arachis"],
        "severity": "high"
    },
    {
        "name": "Fish",
        "keywords": ["fish", "anchovy", "bass", "cod", "salmon", "tuna", "trout", "halibut"],
        "severity": "high"
    },
    {
        "name": "Shellfish",
        "keywords": ["shrimp", "crab", "lobster", "prawn", "crayfish", "clam", "mussel", "oyster", "scallop"],
        "severity": "high"
    },
    {
        "name": "Sesame",
        "keywords": ["sesame", "tahini", "benne"],
        "severity": "medium"
    },
    {
        "name": "Sulfites",
        "keywords": ["sulfite", "sulphite", "sulfur dioxide", "metabisulfite"],
        "severity": "medium"
    }
]

def detect_allergens(ingredient_text):
    """
    Detect allergens in ingredient text
    Returns list of allergen objects with detection status
    """
    text_lower = ingredient_text.lower()
    results = []
    
    for allergen in ALLERGEN_DATABASE:
        matched = []
        detected = False
        
        for keyword in allergen["keywords"]:
            if keyword in text_lower:
                matched.append(keyword)
                detected = True
        
        results.append({
            "name": allergen["name"],
            "keywords": allergen["keywords"],
            "detected": detected,
            "matchedKeywords": matched,
            "severity": allergen["severity"]
        })
    
    return results
